pkg/provisioner/provisioner.go

```diff
@@ -0,0 +1,273 @@
+	Serve(ctx context.Context) error
+	ListPods(ctx context.Context) error
+	DeletePod(ctx context.Context, UID string) error
+}
+
+type k8sProvisioner struct {
+	Namespace  string
+	Clientset  *kubernetes.Clientset
+	Pods       corev1.PodInterface
+	restConfig *rest.Config
+	events     corev1.EventInterface
+}
+
+// NewK8sProvisioner returns a new provisioner implementation that targets
+// Kubernetes using a specific Kubernetes namespace and REST config.
+func NewK8sProvisioner(namespace string, restConfig *rest.Config) (Provisioner, error) {
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	return k8sProvisioner{
+		Namespace:  namespace,
+		Clientset:  clientset,
+		Pods:       clientset.CoreV1().Pods(namespace),
+		restConfig: restConfig,
+		events:     clientset.CoreV1().Events(namespace),
+	}, nil
+}
+
+func (p k8sProvisioner) ListPods(ctx context.Context) error {
+	podList, err := p.Pods.List(ctx, metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/name=wharf-cmd-worker,app.kubernetes.io/managed-by=wharf-cmd-provisioner,wharf.iver.com/instance=prod",
+	})
+	if err != nil {
+		return err
+	}
+	pods := podList.Items
+
+	printPod := func(pod metav1.ObjectMeta, i int) func(logger.Event) logger.Event {
+		return func(ev logger.Event) logger.Event {
+			ev = ev.WithInt("index", i).
+				WithString("UID", string(pod.UID)).
+				WithString("name", pod.Name).
+				WithString("namespace", pod.Namespace)
+
+			for k, v := range pod.Labels {
+				ev = ev.WithString(k, v)
+			}
+
+			return ev
+		}
+	}
+
+	log.Info().WithInt("count", len(pods)).Message("Fetched pods with matching labels.")
+	for i, pod := range pods {
+		log.Info().WithFunc(printPod(pod.ObjectMeta, i)).Message("Pod")
+	}
+
+	return nil
+}
+
+func (p k8sProvisioner) DeletePod(ctx context.Context, UID string) error {
+	podList, err := p.Pods.List(ctx, metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/name=wharf-cmd-worker,app.kubernetes.io/managed-by=wharf-cmd-provisioner,wharf.iver.com/instance=prod",
+	})
+	if err != nil {
+		return err
+	}
+
+	var matchingPod *v1.Pod
+	for _, pod := range podList.Items {
+		if string(pod.ObjectMeta.UID) == UID {
+			matchingPod = &pod
+			break
+		}
+	}
+
+	if matchingPod == nil {
+		return fmt.Errorf("found no pod with appropriate labels matching UID: %s", UID)
+	}
+
+	return p.Pods.Delete(ctx, matchingPod.Name, metav1.DeleteOptions{})
+}
+
+func (p k8sProvisioner) Serve(ctx context.Context) error {
+	podMeta := createPodMeta()
+
+	newPod, err := p.Pods.Create(ctx, &podMeta, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	err = p.waitForInitContainerDone(ctx, newPod.ObjectMeta)
+	if err != nil {
+		return err
+	}
+
+	log.Debug().Message("Init Container done.")
+
+	err = p.waitForAppContainerRunning(ctx, newPod.ObjectMeta)
+	if err != nil {
+		return err
+	}
+
+	log.Debug().Message("App Container running.")
+
+	err = p.streamLogsUntilCompleted(ctx, newPod.Name)
```

Comment:
Provisioner shouldn't wait for a worker to complete nor should it stream its logs. Only the wharf-cmd-worker needs to stream logs from Kuberentes, as even wharf-cmd-aggregator will stream from wharf-cmd-worker and not Kubernetes.

Reply:
Just to clarify, it should still wait until it is running?

Reply:
Actually no. It should make sure it's created, but actually it shouldn't wait for the init container to finish.

When starting a build, it should still just be in the "Scheduling" phase, where it's scheduling the pod, cloning the repo, and so on.

The wharf-cmd-aggregator will wait until it is running/past the init container before marking the build as "Running"